rfcs/text/0008_service_status.md

```diff
@@ -0,0 +1,251 @@
+
+### HTTP responses
+
+As specified in the [_Levels section_](#levels), a service's HTTP endpoints will respond with `503 Unavailable` responses in some status levels.
+
+In both the `fatal` and `unavailable` levels, all of a service's endpoints will return 503s. However, in the `degraded` level, it is up to service authors to decide which endpoints should return a 503. This may be implemented directly in the route handler logic or by using any of the [utilities provided](#status-utilities).
+
+## Status Utilities
+
+Though many plugins should be able to rely on the default status inheritance and associated behaviors, there are common patterns and overrides that some plugins will need. The status service should provide some utilities for these common patterns out-of-the-box.
+
+```ts
+/**
+ * Extension of the main Status API
+ */
+interface StatusSetup {
+  /**
+   * Helpers for expressing status in HTTP routes.
+   */
+  http: {
+    /**
+     * High-order route handler function for wrapping routes with 503 logic based
+     * on a predicate.
+     * 
+     * @remarks
+     * When a 503 is returned, it also includes detailed information from the service's
+     * current `ServiceStatus` including `meta` information.
+     * 
+     * @example
+     * ```ts
+     * router.get(
+     *   { path: '/my-api' }
+     *   unavailableWhen(
+     *     ServiceStatusLevel.degraded,
+     *     async (context, req, res) => {
+     *       return res.ok({ body: 'done' });
+     *     }
+     *   )
+     * )
+     * ```
+     * 
+     * @param predicate When a level is specified, if the plugin's current status
+     *                  level is >= to the severity of the  specified level, route
+     *                  returns a 503. When a function is specified, if that
+     *                  function returns `true`, a 503 is returned.
+     * @param handler The route handler to execute when a 503 is not returned.
+     */
+    unavailableWhen<P, Q, B>(
+      predicate: ServiceStatusLevel |
+        (core: CoreStatus, plugins: Record<string, ServiceStatus>) => boolean,
+      handler: RouteHandler<P, Q, B>
+    ): RouteHandler<P, Q, B>;
+  }
+}
+```
+
+# Drawbacks
+
+1. **The default behaviors and inheritance of statuses may appear to be "magic" to developers who do not read the documentation about how this works.** Compared to the legacy status mechanism, these defaults are much more opinionated and the resulting status is less explicit in plugin code compared to the legacy `mirrorPluginStatus` mechanism.
+2. **The default behaviors and inheritance may not fit real-world status very well.** If many plugins must customize their status in order to opt-out of the defaults, this would be a step backwards from the legacy mechanism.
```

Comment:
Worse case, we could easily add an utility to mimick `mirrorPluginStatus`, right?

Reply:
Sure we could, though it is pretty trivial to do manually (see Basic Example section at top)